db/database.go

```diff
@@ -412,6 +414,10 @@ func NewDatabaseContext(ctx context.Context, dbName string, bucket base.Bucket,
 		UserFunctionTimeout: defaultUserFunctionTimeout,
 	}
 
+	// set up cancellable context based on the background context (context lifecycle should decoupled from the request context)
+	//
+	dbContext.CancelContext, dbContext.cancelContextFunc = context.WithCancel(context.Background())
```

Comment:
I didn't understand this code correctly before and I think we can limit the scope of this PR.

There are two pathways the code cares about:

1. Exiting the changes feed from a blip handler. This is done by `exitChanges` (for takeDBOffline), or the request context cancelling `ChangesOptions.ChangesCtx`, or the database calling `changeListener.Stop` which will close `changeListener.terminator` and then the changes feeds are woken up with `changeListener.tapNotifier.Broadcast` which will return `WaiterCheckTerminated`. This has everything to do with closing a goroutine on the server side and nothing to do with closing a connection from a websocket client. The reason that I thought this would close the connection is that this code eventually exits here https://github.com/couchbase/sync_gateway/blob/08686266a764edab6df8dd0361913af00d60be1d/db/blip_handler.go#L479, but this doesn't shut down the websocket connection.
2. This PR is intended to cover closing the connection on potentially idle clients, and I was definitely confusing the two.

I don't know how to make this more clear, so I'm leaving a lot of text here. I think this isn't important to do now, but `DatabaseContext.exitChanges` would be obsolesced by `DatabaseContext.CancelContext` if we decided to handle the offline case. I like the idea of leaving behind some type of in code bread crumbs, because I don't think we'll prioritize cleaning this code up. I have spent considerable amount of time looking at this code for a past hanging goroutine in https://github.com/couchbase/sync_gateway/commit/7b5a4019ce4834667e8ef891081949f1ec096d80 so I could be suffering from recency bias.

I think a comment as to why you are picking at the start of `NewDatabaseContext` rather than in `StartOnlineProcesses` would be something that seemed not like the typical pattern.

Reply:
Added a comment for clarity.